src/platform/MpiServer.c

```diff
@@ -0,0 +1,611 @@
+
+        if (clientSessionValue && (JSONString == json_value_get_type(clientSessionValue)))
+        {
+            clientSession = (char*)json_value_get_string(clientSessionValue);
+
+            if (IsFullLoggingEnabled())
+            {
+                OsConfigLogInfo(GetPlatformLog(), "Received MpiGetReported request for '%s'", clientSession);
+            }
+
+            mpiStatus = MpiGetReported((MPI_HANDLE)clientSession, (MPI_JSON_STRING*)response, responseSize);
+            status = ((mpiStatus == MPI_OK) ? HTTP_OK : HTTP_INTERNAL_SERVER_ERROR);
+        }
+        else
+        {
+            status = HTTP_BAD_REQUEST;
+        }
+    }
+    else
+    {
+        status = HTTP_BAD_REQUEST;
+    }
+
+    return status;
+}
+
+HTTP_STATUS RouteRequest(const char* uri, const char* request, char** response, int* responseSize)
+{
+    HTTP_STATUS status = HTTP_OK;
+
+    if (IsFullLoggingEnabled())
+    {
+        OsConfigLogInfo(GetPlatformLog(), "HTTP request recieved: %s\n %s", uri, request);
+    }
+
+    if (0 == strcmp(uri, "MpiOpen"))
+    {
+        status = MpiOpenRequest(request, response, responseSize);
+    }
+    else if (0 == strcmp(uri, "MpiClose"))
+    {
+        status = MpiCloseRequest(request, response, responseSize);
+    }
+    else if (0 == strcmp(uri, "MpiSet"))
+    {
+        status = MpiSetRequest(request, response, responseSize);
+    }
+    else if (0 == strcmp(uri, "MpiGet"))
+    {
+        status = MpiGetRequest(request, response, responseSize);
+    }
+    else if (0 == strcmp(uri, "MpiSetDesired"))
+    {
+        status = MpiSetDesiredRequest(request, response, responseSize);
+    }
+    else if (0 == strcmp(uri, "MpiGetReported"))
+    {
+        status = MpiGetReportedRequest(request, response, responseSize);
+    }
+    else
+    {
+        OsConfigLogError(GetPlatformLog(), "%s: invalid request", uri);
+        status = HTTP_NOT_FOUND;
+    }
+
+    if ((status != HTTP_OK) && (status != HTTP_NOT_FOUND) && IsFullLoggingEnabled())
+    {
+        OsConfigLogError(GetPlatformLog(), "Invalid %s request body: %s", uri, request);
+    }
+
+    return status;
+}
+
+static void HandleConnection(int socketHandle)
+{
+    const char* responseFormat = "HTTP/1.1 %d %s\r\nServer: OSConfig\r\nContent-Type: application/json\r\nContent-Length: %d\r\n\r\n%.*s";
+
+    char* uri = NULL;
+    int contentLength = 0;
+    char* requestBody = NULL;
+    HTTP_STATUS status = HTTP_OK;
+    char reasonPhrase[MAX_REASON_PHRASE_LENGTH] = { 0 };
+    char* responseBody = NULL;
+    int responseSize = 0;
+    char* buffer = NULL;
+    int estimatedSize = 0;
+    int actualSize = 0;
+    ssize_t bytes = 0;
+
+    if (NULL == (uri = ReadUriFromSocket(socketHandle, GetPlatformLog())))
+    {
+        OsConfigLogError(GetPlatformLog(), "Failed to read request URI %d", socketHandle);
+        status = HTTP_BAD_REQUEST;
+    }
+
+    if (0 >= (contentLength = ReadHttpContentLengthFromSocket(socketHandle, GetPlatformLog())))
```

Comment:
>0 >=  [](http://example.com/codeflow?start=8&length=5)

Don't need to test for <= 0, just for 0. Look at implementation of ReadHttpContentLengthFromSocket and its unitests. And 0 means not an error, means there is no content. So perhaps change the trace to say that? Not failed to read content, but no content.

Reply:
To add, it is ReadHttpContentLengthFromSocket responsability to log "failed to read HTTP COntent Length").